### DISCUSSION TITLE
Automatic formatter

### DISCUSSION THREAD

**Discussion by kennydo** [apple/pkl#174] [General]
Is there any tooling for automatically formatting a pkl file to some canonical style? I'm looking for something like `dhall format` or `jsonnetfmt`, that I can run as a pre-commit hook to my repo.

I don't care what the style is, as long as it can be applied consistently and automatically.

**Comment by odenix** [comment]
There's the IntelliJ formatter, but I've yet to see a standalone formatter. Probably worth an issue.

**Comment by bioball** [comment]
This is missing, but is something that we'd like to provide. On a similar note, we also want to provide a linter.

**Comment by disrupted** [comment]
it's on their roadmap now https://github.com/apple/pkl/issues/442#issuecomment-2688833372

**Comment by MRDGH2821** [comment]
Can it also sort the mappings/keys?

Like this:

<details><summary> before </summary>
<p>

```pkl
amends "package://github.com/jdx/hk/releases/download/v1.55.0/hk@1.55.0#/Config.pkl"

import "package://github.com/jdx/hk/releases/download/v1.55.0/hk@1.55.0#/Builtins.pkl"

local linters = new Mapping<String, Step> {
  ["betterleaks"] = Builtins.betterleaks
  ["sort-package-json"] = Builtins.sort_package_json
  ["yamlfix"] {
    types = List("yaml")
    check = "yamlfix --check {{files}}"
    fix = "yamlfix {{files}}"
    depends = List("prettier")
  }
  ["yamllint"] = (Builtins.yamllint) {
    depends = List("yamlfix")
  }
  ["tombi"] = Builtins.tombi
  ["mise"] = Builtins.mise
  ["rumdl"] = Builtins.rumdl
  ["rumdl-format"] = Builtins.rumdl_format
  ["typos"] = Builtins.typos
  ["pkl-format"] = Builtins.pkl_format
  ["ls-lint"] {
    check = "ls-lint ."
    batch = false
  }
  ["prettier"] = Builtins.prettier
}

hooks {
  ["pre-commit"] {
    stash = "git"
    fix = true
    steps = linters
  }
  ["commit-msg"] {
    steps {
      ["conventional-commit"] = Builtins.cocogitto_commit_msg
      ["typos"] = Builtins.typos
    }
  }
  ["fix"] {
    fix = true
    steps = linters
  }
  ["check"] {
    steps = linters
  }
}
```

</p>
</details> 

<details><summary> after </summary>
<p>

```pkl
amends "package://github.com/jdx/hk/releases/download/v1.55.0/hk@1.55.0#/Config.pkl"

import "package://github.com/jdx/hk/releases/download/v1.55.0/hk@1.55.0#/Builtins.pkl"

local linters = new Mapping<String, Step> {
  ["betterleaks"] = Builtins.betterleaks
  ["ls-lint"] {
    check = "ls-lint ."
    batch = false
  }
  ["mise"] = Builtins.mise
  ["pkl-format"] = Builtins.pkl_format
  ["prettier"] = Builtins.prettier
  ["rumdl"] = Builtins.rumdl
  ["rumdl-format"] = Builtins.rumdl_format
  ["sort-package-json"] = Builtins.sort_package_json
  ["tombi"] = Builtins.tombi
  ["typos"] = Builtins.typos
  ["yamlfix"] {
    types = List("yaml")
    check = "yamlfix --check {{files}}"
    fix = "yamlfix {{files}}"
    depends = List("prettier")
  }
  ["yamllint"] = (Builtins.yamllint) {
    depends = List("yamlfix")
  }
}

hooks {
  ["check"] {
    steps = linters
  }
  ["commit-msg"] {
    steps {
      ["conventional-commit"] = Builtins.cocogitto_commit_msg
      ["typos"] = Builtins.typos
    }
  }
  ["fix"] {
    fix = true
    steps = linters
  }
  ["pre-commit"] {
    fix = true
    stash = "git"
    steps = linters
  }
}

``` 

</p>
</details>

**Reply by HT154** [reply]
Mapping key order matters in Pkl, so that code is not equivalent!

Per the [language reference](https://pkl-lang.org/main/current/language-reference/index.html#mappings):
> A value of type `Mapping` is an **ordered** collection of values indexed by key.

While it wouldn't make sense for the language's canonical formatter to make this type of change, the upcoming `pkl:syntax` stdlib module could be used to build a separate tool that does this in cases where it is safe and possible.

It's also worth noting that a Mapping key may be an arbitrary expression, not just a String or Int literal, so in many cases Pkl evaluation (and a custom comparator) may be required for sorting.

**Reply by MRDGH2821** [reply]
Fair point :thinking: 

Is it the same in case of `hooks` section in the example I put?

**Reply by HT154** [reply]
From a cursory glance, the entry order doesn't seem to matter for hk specifically, but the part about expressions still applies: you could use an expression (e.g. reading a local property) as a key there:
```pkl
local myHook = "check"
hooks {
  [myHook] {
    steps = linters
  }
}

```
